pkg/controller/elasticsearch/driver/upgrade_deletion_driver.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Delete runs through a list of potential candidates and select the ones that can be deleted.
+// Do not run this function unless driver expectations are met.
+func (ctx *rollingUpgradeCtx) Delete() ([]corev1.Pod, error) {
+	if len(ctx.podsToUpgrade) == 0 {
+		return nil, nil
+	}
+
+	// Check if we are not over disruption budget
+	// Upscale is done, we should have the required number of Pods
+	statefulSets, err := sset.RetrieveActualStatefulSets(ctx.client, k8s.ExtractNamespacedName(&ctx.ES))
+	if err != nil {
+		return nil, err
+	}
+	expectedPods := statefulSets.PodNames()
+	unhealthyPods := len(expectedPods) - len(ctx.healthyPods)
+	maxUnavailable := 1
+	// TODO: use GroupingDefinition
+	if ctx.ES.Spec.UpdateStrategy.ChangeBudget != nil {
+		maxUnavailable = ctx.ES.Spec.UpdateStrategy.ChangeBudget.MaxUnavailable
+	}
+	allowedDeletions := maxUnavailable - unhealthyPods
+	// If maxUnavailable is reached the deletion driver still allows one unhealthy Pod to be restarted.
+	// TODO: Should we make the difference between MaxUnavailable and MaxConcurrentRestarting ?
+	maxUnavailableReached := (maxUnavailable - unhealthyPods) <= 0
+
+	// Step 2. Sort the Pods to get the ones with the higher priority
+	candidates := make([]corev1.Pod, len(ctx.podsToUpgrade)) // work on a copy in order to have no side effect
+	copy(candidates, ctx.podsToUpgrade)
+	sortCandidates(candidates)
+
+	// Step 3: Apply predicates
+	predicateContext := NewPredicateContext(
+		ctx.esState,
+		ctx.healthyPods,
+		ctx.podsToUpgrade,
+		ctx.expectedMasters,
+	)
+	deletedPods, err := applyPredicates(predicateContext, candidates, maxUnavailableReached, allowedDeletions)
```

Comment:
Naming nit: `deletedPods` indicates that the pods have already been deleted which is not the case iiuc